### DISCUSSION TITLE
Sub-daily budget failed

### DISCUSSION THREAD

**Discussion by Kutcha7** [EDmodel/ED2#404] [General]
Hi everyone, 
I tried to start a simulation and got the error sub daily budget failed. I checked the log file for NaNs or anything that could lead me to the error but I can't really find the source. Note that I only get this error when I use the opt executable. When I use the dbg it stops after initialising the met cycle and the error I get is related to "call_sdh5_open_f".

I've attached my logfile and the met header. I appreciate any feedback and happy to share more files! 

1
/sci/labs/efratshef/irobk/Yatir/Future/gfdl-esm4/Weather/ISIMP_Israel_gfdl-esm4_ssp585_
2 2 0.5 0.5 35.065668 31.342226
13
   'hgt'   'tmp'  'pres'    'sh'  'ugrd'  'vgrd' 'prate' 'dlwrf' 'nbdsf' 'nddsf' 'vbdsf' 'vddsf' 'co2' 
27 3600 3600 3600 3600 3600 3600 3600 3600 3600 3600 3600 3600
 4 1 1 1 1 1 0 1 1 1 1 1 1

Thanks! 
Katja 
[output.txt](https://github.com/user-attachments/files/16706798/output.txt)

**Comment by mdietze** [comment]
Personally, I'd prioritize the error that shows up under debug model. What's the full trace of the error? I don't know what call_sdh5_open_f is but I'm guessing a low level HDF5 function (i.e. not our call) so you need to know where it's being called from and whether the file being open is present, valid, complete, and has the correct units. Misread met data could easily be the source of your downstream budget issue.

**Comment by mpaiao** [comment]
Most met driver issues typically pop up during sunrise and sunset, because this one is happening at midnight on the second day, so my hunch is that this is somehow a bug related to phenology. I see you started the run in January, which would be winter at your site, as a suggestion, maybe try setting the initial time in May or June. If the model goes through and keeps running, it may be indication of an issue with phenology.

**Comment by mpaiao** [comment]
But to @mdietze's point, met drivers are often the source of crashes. The main cause of problems is when IMETAVG is not properly set for the site, so do double check that variable too. It's on my to do list, but eventually I want to migrate IMETAVG to the met driver header (so all met driver info is in a single place), and after that I want to implement a refraction correction for sunset/twilight that should reduce/eliminate the problems, I have a prototype of it in R, just need to migrate it to ED2.

**Reply by Kutcha7** [reply]
Hi Mike and Marcos, 
thanks for the feedback so far! This problem only occurred when I try to ran an initialised vegetation simulation with future climate data. 
- I checked if it makes a difference to start the simulation in May -> I get the same error 
- I checked the met forcing again for anything unusual and compared them with the historical data that I used for the potential vegetation simulation -> I couldn't find any NA's or unusual values (I implemented many checks when I created the data so it shouldn't have any faulty values) 
- I tried to run the initialised vegetation simulation with the historical met forcings that worked for the potential vegetation simulation, and I got the same error 

So my hunch is that the error is related to the .pss .css file. I'm not really sure what to look for. Do you have any suggestions? 

Thanks again!

**Reply by Kutcha7** [reply]
> But to @mdietze's point, met drivers are often the source of crashes. The main cause of problems is when IMETAVG is not properly set for the site, so do double check that variable too. It's on my to do list, but eventually I want to migrate IMETAVG to the met driver header (so all met driver info is in a single place), and after that I want to implement a refraction correction for sunset/twilight that should reduce/eliminate the problems, I have a prototype of it in R, just need to migrate it to ED2.

I set IMETAVG to 1 so it should be fine right?

**Comment by Kutcha7** [comment]
I still haven't been able to run it and get a sub-daily budget fail related to carbon budget. 

So far I have: 
- replaced/played around with some of the inputs  (vegetation parameters, met forcings, .pss file) 
- changed the disturbance type (trk) 
- made the model start at different times of the year 
- printed out all carbon related outputs in the log file 

but I can't find the source.... 

Does anyone have any suggestions what else I could try? 

Attached my log file, pss and css (as csv files). 

Any help is appreciated! 

[output_StartMay15_withCpools.txt](https://github.com/user-attachments/files/16836625/output_StartMay15_withCpools.txt)
[Yat.lat31.342226lon35.065668css.csv](https://github.com/user-attachments/files/16836685/Yat.lat31.342226lon35.065668css.csv)
[Yat.lat31.342226lon35.065668pss.csv](https://github.com/user-attachments/files/16836686/Yat.lat31.342226lon35.065668pss.csv)

**Comment by mpaiao** [comment]
@Kutcha7 regarding the IMETAVG, it depends on what the time stamp means on your particular meteorological driver. If the time stamp corresponds to the end of the averaging period, then yes, it should be 1. 

Your error is unlikely to be caused by IMETAVG, though. I see that you are only using PFT 7, which is an evergreen PFT, so I doubt the issue is phenology. Because the crash is happening on the second day, I think it may be a lingering problem with the initial conditions that is only checked on the second day (if I remember correctly, the CARBON block is only checked at daily time steps).

I noticed in your css file that the "cohort" column has repeated letters for multiple cohorts in the same patch. I would try to keep them unique within each patch, but I don't remember if this would cause problems in the initialisation. If that doesn't solve the problem, try running the model with IED_INIT_MODE=0 (near bare ground). This will ignore the pss/css files and start with a minimum number of plants. If the model runs fine, then I think it's safe to say that the issue is indeed in the pss/css files.

**Reply by Kutcha7** [reply]
I agree that I don't think it's related to phenology... 

Based on your recommendations, I changed the cohort identifier to be unique per patch, but it still resulted in carbon budget fail. 
Starting the simulation with init_mode = 0 is does not fail and runs until the end of the climate series. As it has before with historical climate.. 

However, these simulations are not very meaningful as this forest was planted in the desert and had to be irrigated in the first years. Consequently, the simulations starting from seedlings basically show what would have happened without human intervention -> the forest never fully develops under these very dry conditions and only very few small seedlings "survive". 

Is there anything else in the initial conditions that can be changed to avoid carbon budget fail? 
Would you say this is rather related to the vegetation information or patch?
Or would you recommend to try to simulate the plantation + irrigation? 

Thanks again! :-)

**Reply by mpaiao** [reply]
I was suggesting IED_INIT_MODE=0 just to check whether the problem might have been with initial conditions or something else, not as an alternative set up for the science runs. Let me check the pss/css files once again and see if I can think of anything that may be triggering the error.

To clarify, are you currently running the model with irrigation? I would be surprised that irrigation would cause issues to carbon (as opposed to water or energy), but if you are using irrigation, maybe try without it just to try to isolate the problem? Also, were you able to run the model with historical climate and IED_INIT_MODE=6?

**Reply by Kutcha7** [reply]
Fair point. And yes it seems to be connected to initial conditions.

Not not running with irrigation. Didn't even know it was an option :) 

And no historical climate and INIT_MODE = 6 did also result in carbon budget fail unfortunately. 

Thanks again and let me know if I can provide any additional files that might be helpful!

**Reply by mpaiao** [reply]
Would you mind describing a bit more the forest inventory sampling protocol at Yatir? I am asking it because I see a few things in your pss/css files that may need adjustment, but I can probably advise better based on how the measurements were made.

- **pss file**. I see all areas are set to 0.004, which I assume are the areas of the plots in ha? I think ED can handle this, but technically the areas should be relative to the total area sampled, so in your case they would become 0.05 (1/20 since all patches have the same area).
- **css file**. I see that n varies across cohorts, and is smaller for small cohorts, which is a bit unusual. If you are using forest inventory plots to generate the files, the easiest way is to list every individual tree and treat them as a "single-tree cohort". In this case, the value of n should always be 1/[sampled area in that plot in m<sup>2</sup>]. Let's say, if the plot area was indeed 0.004 ha (40 m<sup>2</sup>) and there was no size-dependent sampling effort (i.e., trees with DBH ≥ 5 cm are measured everywhere in the plot), then n should be 0.025 m<sup>&minus;2</sup> for every tree. But if there was sub-sampling, say trees with 5 ≤ DBH < 10 cm were measured only in a 4m<sup>2</sup> sub-plot, then n for these smaller trees should be 0.25 m<sup>&minus;2</sup>.

My concern now is that many trees are being rejected due to low stem number density, and that is adding up to the lack of carbon balance, though ED2 should be able to handle this, so there is likely a latent bug in the code.

**Reply by Kutcha7** [reply]
Sure! 
20 plots were measured, each approximately 5 Dunam (~5000m^2). In each plot, they measured all individuals and classified them in different health categories (4 categories from dead to very healthy). Important to note here is that I did not include dead trees, tree stumps or trees that were barely alive). 

pss file: Oopsie, I think it was a coding error. I will correct this! Thanks for spotting this! 

css file: Not sure I understood correctly what n is, I thought that it is the number of individuals per cohort per plot area in m^2, as a proxy of density. Meaning if there are 5 individuals in cohort 1 and the plot area is 3500 m^2 y it’s 5/3500. But that was probably wrong then? 

So it could be a problem of not including the dead trees/unhealthy trees and my probably incorrect N causing a too low density? As well as defining that every 5cm dbh change is a new cohort?  

Thanks again!

**Reply by mpaiao** [reply]
I am curious if the patch area is what is causing the error then.

If your first cohort has 5 individuals and the plot area is 3500 m<sup>2</sup>, then yes, n = 5/3500. In general, I don't even bother creating cohorts before sending the information to ED2, I just list every tree with their exact DBH and set n = 1 / [sampled  area in the plot]. ED2 can and will fuse the trees into cohorts if needed. 

Definitely do not add dead trees as cohorts, although you may want to use that information to set the litter pools. The code will run fine if you don't use it, but beware that heterotrophic respiration may become biased (maybe less of an issue there because it is very dry). 

My concern with very small density is that below a certain threshold, the code terminates the cohort, but I don't think this should happen with a plot of about 5000 m<sup>2</sup>. That said, if you have the exact location of the trees within the plot, and the plot can be easily broken into smaller plots, maybe do generate more patches, it may capture more heterogeneity this way. In the Amazon, I normally split plots into subplots of about 400m<sup>2</sup>.

**Reply by Kutcha7** [reply]
I made several adjustments to n and let the model create cohorts itself. First it just resulted in the model crashing a few years later but I adjusted it further which seems to have fixed the problem. I have a full model run now! Thank you so much!

**Reply by mpaiao** [reply]
Great to hear it worked at last. Out of curiosity, which adjustments did you do? I am asking because it could be useful for future users. Thanks!

**Reply by mpaiao** [reply]
By the way, I never replied your question on trees that are barely alive. Currently ED does not have an explicit way of telling which trees are healthy and which ones are badly damaged or otherwise in poor health, though it would be a very nice addition to the model for anyone interested in implementing it ;).

If you do add them, just beware that they will be just like the healthy ones.

**Reply by Kutcha7** [reply]
Agree I also think it would be a nice addition. For now I have other priorities (adapting ED to drylands) but I will keep it in mind :) 

After you asked me about the sample design, I asked the people who did the survey again and they told me that they had subplots of 8m^2. So I adjust n in the .css homogeneously to 1/8. It still resulted in a C budget fail.. so I assumed there might have been a size dependent sample effort and implemented that all trees with dbh < 10 -> n = 1/4. So here, I only adjust n and  did not change the number of patches. With these settings I got a whole simulation run without problems. 
However, then I let the same settings run with climate projections from a different climate model (fairly similar but more rain), and it resulted in a C budget fail after 4 years of simulation. So this time, I implemented more patches and changed n back to 1/8 for all entries. And I got a full run again.. Not sure what exactly causes this bug but it's definitely a strange one (and not intuitive to solve).

**Reply by Kutcha7** [reply]
Actually depending on climate projection it still keeps failing after a few years of simulation.. Whenever I decrease the density or include more patches, it runs for a little bit longer. I'm not sure what the best practice here is. Or what really causes the carbon budget fail. 
Could it be related to sth else, such as decrease in LAI, critical height? The PFT is not properly calibrated yet so many individuals suffer under the harsh climate conditions there.
